app/org/sagebionetworks/bridge/models/schedules/ScheduleContext.java

```diff
@@ -200,13 +228,22 @@ public Builder withContext(ScheduleContext context) {
         }
         
         public ScheduleContext build() {
-            // pretty much everything else is optional. I would like healthCode to be required, but it's not:
-            // we use these selection criteria to select subpopulations on sign up.
-            if (now == null) {
-                now = (initialTimeZone == null) ? DateTime.now() : DateTime.now(initialTimeZone);
+            // ScheduleContext is validated to have the correct values:
+            // - requestTimeZone is required
+            // - now is advisable, but still not required
+            // - endsOn OR daysAhead needs to be supplied to calculate endsOn from now
+            if (initialTimeZone == null) {
+                initialTimeZone = requestTimeZone;
+            }
+            if (now == null && requestTimeZone != null) {
+                now = DateTime.now(requestTimeZone);
+            }
+            if (endsOn == null && now != null) {
+                endsOn = now.plusDays(daysAhead).withHourOfDay(23).withMinuteOfHour(59).withSecondOfMinute(59)
+                        .withMillisOfSecond(999);
```

Comment:
now and requestedTimeZone may be different. This may lead you into weird states where endsOn isn't in the requestedTimeZone.